AlphaWallet/Core/Coordinators/WalletBalanceCoordinator.swift

```diff
@@ -0,0 +1,297 @@
+
+                balances.insert(tokenObject)
+            }
+
+            walletBalanceSubscribable.value = .init(wallet: wallet, values: balances)
+
+            strongSelf.notifyWalletSummary()
+        }
+
+        //NOTE: for each wallet we handle for adding or deletion event and fire fetch coin ticker for all available tokens
+        pair.tokensUpdateSubscibableKey = fetcher.tokensChangeSubscribable.subscribe { [weak self] _ in
+            guard let strongSelf = self else { return }
+            
+            DispatchQueue.main.async {
+                //NOTE: grabbing for tokens should be on main queue
+                strongSelf.fetchTokenPrices()
+            }
+        }
+
+        return pair
+    }
+
+    func subscribableBalance(wallet: Wallet) -> Subscribable<WalletBalance> {
+        if let pair = balanceFetchersPairs[wallet] {
+            return pair.balanceSubscribable
+        } else {
+            let pair = createWalletBalanceFetcherAndSubscriptionPair(wallet: wallet)
+            pair.balanceFetcher.start()
+
+            balanceFetchersPairs[wallet] = pair
+
+            return pair.balanceSubscribable
+        }
+    }
+
+    private func fetchTokenPrices() {
+        firstly {
+            coinTickersFetcher.fetchPrices(forTokens: availableTokenObjects)
+        }.done(on: .main, { [weak self] tickers in
+            guard let strongSelf = self else { return }
+
+            strongSelf.updateSubscribableWithTickers(tickers: tickers)
+            strongSelf.tickers = tickers
+        }).cauterize()
+    }
+
+    private func updateSubscribableWithTickers(tickers: [AddressAndRPCServer: CoinTicker]) {
+        let availableFetchers = balanceFetchersPairs.mapValues { $0.balanceFetcher }
+
+        for fetcher in availableFetchers {
+            var balances = Set<Activity.AssignedToken>()
+
+            for var tokenObject in fetcher.value.tokenObjects {
+                tokenObject.ticker = tickers[tokenObject.addressAndRPCServer]
+
+                balances.insert(tokenObject)
+            }
+            
+            balanceFetchersPairs[fetcher.key]?.balanceSubscribable.value = .init(wallet: fetcher.key, values: balances)
+        }
+
+        notifyWalletSummary()
+    }
+
+    private func notifyWalletSummary() {
+        if let value = subscribableWalletsSummary.value {
+            if value == walletSummary {
+                //no-op
+            } else {
+                subscribableWalletsSummary.value = walletSummary
+            }
+        } else {
+            subscribableWalletsSummary.value = walletSummary
+        }
+    }
+
+    private var walletSummary: WalletSummary {
+        return .init(values: balanceFetchersPairs)
+    }
+}
+
+extension Wallet: Hashable {
+    func hash(into hasher: inout Hasher) {
+        hasher.combine(address.eip55String)
+    }
+}
+
+private class WalletBalanceFetcher: NSObject {
+    private static let updateBalanceInterval: TimeInterval = 15
```

Comment:
Shall we extend this to a larger value, maybe 60s? Worried that it will blow the limits if there are many wallets with many tokens.

Reply:
sure we can do that. as all actions performs in background it shouldn't load the CP too mutch.